### ISSUE TITLE
Failed to decode dedup "gob: wrong type (int) for received field .Port"

### ISSUE THREAD

**Issue by erulabs** [hashicorp/consul-template#920]
### Consul Template version

v0.18.2 and v0.18.1 with deduplication on. Note that this issue _does not happen_ with deduplicaton off.

### Configuration

```liquid
{"servers": [
        {{range $index, $service := service "memcached"}}{{if gt $index 0}}, {{end}}"{{.Address}}:{{.Port}}"{{end}}
      ]
}
```

### Command

```shell
consul-template -consul localhost:8500 -dedup -syslog -template="/etc/mcrouter/mcrouter.ctmpl:/etc/mcrouter/mcrouter.conf:service mcrouter restart"
```

### Debug output
```
2017/03/31 18:49:44.426580 [INFO] (dedup) attempting lock for template hash 630aa26fad72d6f5faa6ceea46d38254
:49:44.409315 [INFO] (dedup) attempting to create session
2017/03/31 18:49:44.409792 [INFO] (dedup) starting watch for template hash 630aa26fad72d6f5faa6ceea46d38254
2017/03/31 18:49:44.409875 [INFO] (dedup) listing data for template hash 630aa26fad72d6f5faa6ceea46d38254
2017/03/31 18:49:44.411945 [ERR] (dedup) failed to decode 'consul-template/dedup/630aa26fad72d6f5faa6ceea46d38254/data': gob: wrong type (int) for received field .Port
```

### Expected behavior
Pre 0.18.1 (the last version we were using was 0.15.0), the .Port field would be cast properly as a string (even tho is configured on consul as a int) with no errors. I would not expect consul to fail to coerce a int to a string (keep in mind we're writing to a file... at the end of the day these are _always_ strings and there is no risk to simply printing/coercing the value)

### Actual behavior
Template is not rendered and an error is encountered:
```
2017/03/31 17:57:18.100862 [ERR] (dedup) failed to decode 'consul-template/dedup/630aa26fad72d6f5faa6ceea46d38254/data': gob: wrong type (int) for received field .Port
```

### Steps to reproduce

1. Turn on deduplication
2. Attempt to cast int to string in consul template

For now I have downgraded to 0.15 and the issue is resolved and my templates are rendering as expected.

Thanks!

**Comment by sethvargo** [comment]
Hi @erulabs 

What version of Consul are you using?

**Comment by erulabs** [comment]
This occurs on 0.18.1 and 0.18.2 - does not occur on 0.15 as far as I can tell (still verifying that)

**Comment by erulabs** [comment]
Oh I apologize @sethvargo - consul version is 0.7.4

**Comment by sethvargo** [comment]
I wonder if the types changed at some point, and you have old data. Have you tried clearing that key manually in Consul's KV store? The gob stuff is very internal to Go from what I understand. I know one thing that changed between those versions is the version of Go we compile with.

**Comment by erulabs** [comment]
@sethvargo Very nice - good call!

The command to fix was:
```
consul kv delete consul-template/dedup/630aa26fad72d6f5faa6ceea46d38254/data
```

After that consul-template appears to be good to go! Potentially worth some sort of system that deletes the deduplicated kv if it fails to be parsed?

**Comment by erulabs** [comment]
@sethvargo I believe I found the source of this issue. Removing the KV worked briefly, but the source of the problem appears to be that deduplicated templates between different versions of consul-template do not play well.

A deduped k/v from 0.15.0 is not able to be loaded by a 0.18.1+ consul-template instance.

Perhaps consul-template ought to keep deduped templates per-version?

Downgrading entirely to 0.15 seems to work for my production environment - in my development environment upgrading entirely to 0.18.2 also seems to resolve the issue (note that I had to purge the K/V dedupe store after upgrading to get things playing nicely)
